protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,509 @@
+                Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                    upgrade,
+                    local_addr,
+                    remote_addr,
+                }))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                        upgrade: RelayedListenerUpgrade::Inner(upgrade),
+                        local_addr,
+                        remote_addr,
+                    })))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr))))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::Error(err)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Error(EitherError::A(err)))))
+                }
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => {}
+            },
+            RelayListenerProj::Relayed {
+                queued_new_addresses,
+                from_behaviour,
+                msg_to_behaviour,
+            } => {
+                if let Some(msg) = msg_to_behaviour {
+                    match Future::poll(msg.as_mut(), cx) {
+                        Poll::Ready(Ok(())) => *msg_to_behaviour = None,
+                        Poll::Ready(Err(e)) => {
+                            return Poll::Ready(Some(Err(EitherError::B(e.into()))))
+                        }
+                        Poll::Pending => {}
+                    }
+                }
+
+                if let Some(addr) = queued_new_addresses.pop_front() {
+                    return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))));
+                }
+
+                match from_behaviour.poll_next_unpin(cx) {
+                    Poll::Ready(Some(ToListenerMsg::IncomingRelayedConnection {
+                        stream,
+                        src_peer_id,
+                        relay_addr,
+                        relay_peer_id: _,
+                    })) => {
+                        return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                            upgrade: RelayedListenerUpgrade::Relayed(Some(stream)),
+                            local_addr: relay_addr.with(Protocol::P2pCircuit),
+                            remote_addr: Protocol::P2p(src_peer_id.into()).into(),
+                        })));
+                    }
+                    Poll::Ready(Some(ToListenerMsg::Reservation(Ok(Reservation { addrs })))) => {
+                        let mut iter = addrs.into_iter();
+                        let first = iter.next();
+                        queued_new_addresses.extend(iter);
+                        if let Some(addr) = first {
```

Comment:
Why prioritize the addrs[0], hence break the chronology, instead of reporting them like this:
```suggestion
                        queued_new_addresses.extend(addrs);
                        if let Some(addr) = queued_new_addresses.pop_front() {
```

Reply:
My code here is confusing. At this point in time `queued_new_addresses` is guaranteed to be empty due to the previous:

```rust
            if let Some(addr) = self.queued_new_addresses.pop_front() {
                return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))));
            }
```

Thus chronology is not broken.

In addition to being confusing, it might as well return `Poll::Pending` without registering a `Waker` in the case where `addrs` is empty.

Both of the above are fixed with daa06865. @elenaf9 what do you think? Is this more intuitive? Am I missing something else?